### DISCUSSION TITLE
Geofenced custom maps

### DISCUSSION THREAD

**Discussion by mattghub1** [time4tea/gopro-dashboard-overlay#132] [Ideas]
Wondering if it's possible to implement a component to allow custom maps. By custom an image file geo references somehow. There's a few tools to help prep this like quickroute. This would support sports like orienteering where a custom map is used and the gps gets overlayed. 

An example used https://youtu.be/d4A4NBu4UiY which is created from orientview. The software zooms/overlays the map then rotates and split intervals. It's great software but sadly going out of data and some limitations where's your software doesn't have these and performs extremely well. 

Also happy to try and help but struggling to follow the code just now but I need to spend a bit more time around the component redraw element and helper functions.

**Comment by time4tea** [comment]
Hi - thanks for sharing this use-case which was one I hadn't really thought about at all. I read the site for "orientview" very quickly - https://github.com/mikoro/orientview - and if I understand correctly, it allows to use a big TIFF file as a map. The map will have all sorts of orienteering-specific markings on it.
Just throwing some random thoughts around, I was wondering if it might be "good enough" to create a special map tile server that could serve map tiles from this TIFF file (or some other file format of course). You could start it up something like "image-tile-server --bbox [some gps bounding box] file.tiff ... then by pointing the gopro-dashboard at this local map tile server (something that you cant currently do, but could be easy to add) then the tiles could come from the TIFF image, and the rest of the dashboard software would work as normal. We'd need to be able to force gopro-dashboard to not cache the tiles on disk - as they would change all the time, but thats easy enough.
Would that be useful for some part of the stuff you'd like to add? Is there anything else? Also, perhaps I completely misunderstood! Please do correct me if thats the case...

**Comment by time4tea** [comment]
ok - I have a rough and ready solution.... It might be "good enough" - there are a few steps to do this right now which we can tidy up in later releases.

1 - get your orienteering map, in as large a scan as possible. I found an example of one - i think - on the web - https://www.bristolorienteering.org.uk/sites/default/files/inline-images/Screenshot%202020-06-03%20at%2015.37.15.png
2 - get maptiler engine - https://www.maptiler.com/engine/download/
3 - use maptiler engine to create map tiles from your map - open the image, geolocate it, using two points. then export. export settings should be the defaults, but use *png* format. pick an empty folder on your computer to export to. this will create a load of files.
4 - in a terminal window, go to the folder you used for the export, and use `python3 -mhttp.server` - this will set up a simple webserver to serve your tiles. This will run on port 8000.
5 - using your GPX file, or your gopro video, use `gopro-overlay` with `--map-style local` 

You will end up with a map rendered using your local tiles!

Here is a screenshot from a video I made using the above image and approach:

![image](https://github.com/time4tea/gopro-dashboard-overlay/assets/395841/c752232a-c125-4f2d-ab1b-ae91e40cff8e)

`--map-style local` will be in version 0.97.0 whihc I will release today.

You may find that a larger map and/or a higher zoom works better - to do this use a custom layout.

**Reply by time4tea** [reply]
You will get a lot of http errors for places where the tiles don't exist... but eventually this will settle down, and it should work ok in the end. If this is useful, I'll streamline the workflow a bit and make sure the errors are not so many...

**Reply by time4tea** [reply]
version 0.97.0 is now available

**Reply by mattghub1** [reply]
this looks amazing. Will give it a try right away. Didnt even think about a map tile server. Thank you very much. Honestly didnt expect it to be coded in (was going to attempt a pull req once i got my head round the code) let alone this quickly amazing.

**Reply by time4tea** [reply]
you will have to sign up for maptiler to do this... but only the free account.. later if this is a good approach we can include a tileserver in this program that can use a geolocated image - but this allows to explore the concept without spending much time to make it work.

**Reply by time4tea** [reply]
@mattghub1 did you have any opportunity to give this a go? Of course no obligation! If you did, did it work to some extent & Would it be worth building into a more robust feature? Happy to hear even if negative results... thanks!

**Reply by mattghub1** [reply]
Got it working late last night well sort of. Using the defaults I got the below. I'll need to scale up the image and figure out what happened to the top map. 

Not sure if its of any use but would a kmz file help? Quickroute gives an option to export as a kmz including replay and the georeferneced image. Certainly helped with the coordinates for map tiler. 

![image](https://github.com/time4tea/gopro-dashboard-overlay/assets/14833046/dba08fe6-169a-47bc-95fa-78ea3a053379)

**Reply by time4tea** [reply]
I think the tiles in the top map are saying "map tile missing" or something..  but by the looks of it it its sort of working... I dontquite understand how the bottom map works but not the top. There is also that white band in the bottom map? Thanks for giving it a go. I think it shows some promise, but not quite there yet..

Would you be able to send me the map image and either gopro movie or gpx file? This might help me understand a bit more about the particular set of files you're using. If not - totally no problem. All files are stored on encrypted drives. Ifits OK, send using wetransfer or Google drive link or something to gopro-overlay [at] time4tea.net

In general, it's better to generate the overlay directly from the gopro, rather than via an intermediate format, as the timing is super-exact.

**Reply by time4tea** [reply]
It would also be really useful to understand your current workflow a little..which files get used by which programs etc.. if you have the time to explain. Thanks!!

**Reply by mattghub1** [reply]
Of course no problem at all. I'll start an upload just now and drop you an email with the details. I'll be running a second event tomorrow so will have another example dataset.

**Reply by pialanrveze** [reply]
> Got it working late last night well sort of. Using the defaults I got the below. I'll need to scale up the image and figure out what happened to the top map.
> 
> Not sure if its of any use but would a kmz file help? Quickroute gives an option to export as a kmz including replay and the georeferneced image. Certainly helped with the coordinates for map tiler.
> 
> ![image](https://private-user-images.githubusercontent.com/14833046/242080335-dba08fe6-169a-47bc-95fa-78ea3a053379.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjczNDQ5NTcsIm5iZiI6MTcyNzM0NDY1NywicGF0aCI6Ii8xNDgzMzA0Ni8yNDIwODAzMzUtZGJhMDhmZTYtMTY5YS00N2JjLTk1ZmEtNzhlYTNhMDUzMzc5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI2VDA5NTczN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYzNDI1ZTAwYzAxMjFhYTllMjg2MmQyYTkzNDIxYjVlMjcxZmM0NmRlMTE3ZGIzNmI3OWJkNjczNmU2ZjRjYjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.3F1M8En8bu9eHTwS34zWD0G5SAO3wFqQm0Tw2r0iaXA)

Stuck with the same problem, set zoom to 15 in layout and it is working now.
